examples/knative/go/kn-go-tagging/tagging/tagging_test.go

```diff
@@ -0,0 +1,396 @@
+	"io/ioutil"
+	"net/url"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/cloudevents/sdk-go/v2/event"
+	"github.com/vmware/govmomi/simulator"
+	_ "github.com/vmware/govmomi/vapi/simulator"
+	"github.com/vmware/govmomi/vapi/tags"
+	"github.com/vmware/govmomi/vim25"
+	"github.com/vmware/govmomi/vim25/mo"
+	"github.com/vmware/govmomi/vim25/types"
+	"go.uber.org/zap/zaptest"
+	"knative.dev/pkg/logging"
+)
+
+func TestParseEventMoRef(t *testing.T) {
+	type vm struct {
+		vmType string
+		Value  string
+	}
+
+	var tests = []struct {
+		name      string
+		path      string
+		expectErr bool
+		want      *vm
+	}{
+		{
+			"Test that event is readable",
+			"testdata/event.json",
+			false,
+			&vm{vmType: "VirtualMachine", Value: "vm-10000"},
+		},
+		{
+			"Event should return error if vm type and value are null",
+			"testdata/eventErr1.json",
+			true,
+			nil,
+		},
+		{
+			"Event should return error if vm info is null",
+			"testdata/eventErr2.json",
+			true,
+			nil,
+		},
+		{
+			"Event should return error if vm parent is null",
+			"testdata/eventErr3.json",
+			true,
+			nil,
+		},
+		{
+			"Event should return error if data is null",
+			"testdata/eventErr4.json",
+			true,
+			nil,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			e := jsonFileToEvent(tt.path)
+			moRef, err := parseEventMoRef(e)
+
+			if tt.expectErr && err == nil {
+				t.Error("expected error, but got none")
+			}
+
+			if !tt.expectErr && err != nil {
+				if moRef.Type != tt.want.vmType {
+					t.Errorf("got: %q, want: %q", moRef.Type, tt.want.vmType)
+				}
+
+				if moRef.Value != tt.want.Value {
+					t.Errorf("got: %q, want: %q", moRef.Value, tt.want.Value)
+				}
+			}
+		})
+	}
+}
+
+func TestRun(t *testing.T) {
+	const (
+		username = "administrator@vsphere.local"
+		password = "pass"
+	)
+
+	secretsDir := createSecret(t, username, password)
+	defaultEnv := EnvConfig{
+		Port:       50001,
+		SecretPath: secretsDir,
+		Insecure:   true, // vcsim
+	}
+
+	t.Run("fail with authentication error", func(t *testing.T) {
+		model := simulator.VPX()
+		defer model.Remove()
+		if err := model.Create(); err != nil {
+			t.Fatalf("create vcsim model: %v", err)
+		}
+		model.Service.Listen = &url.URL{
+			User: url.UserPassword("not-my-username", password),
+		}
+
+		simulator.Run(func(ctx context.Context, client *vim25.Client) error {
+			defaultEnv.VCAddress = client.URL().String()
+			if err := setEnv(defaultEnv); err != nil {
+				t.Fatalf("set environment variables: %v", err)
+			}
+
+			ctx, cancel := context.WithCancel(ctx)
+			defer cancel()
+			ctx = logging.WithLogger(ctx, zaptest.NewLogger(t).Sugar())
+
+			go func() {
+				<-time.After(time.Second)
+				logging.FromContext(ctx).Debug("stop test server")
+				cancel()
+			}()
+
+			_, err := NewClient(ctx)
+			if err != nil {
+				msg := "SOAP session manager login: ServerFaultCode: Login failure"
+				if !strings.Contains(err.Error(), msg) {
```

Comment:
nit: use `assert.Error` to simplify test assertions (https://pkg.go.dev/github.com/vieux/docker@v0.6.3-0.20161004191708-e097c2a938c7/pkg/testutil/assert#Error)